### ISSUE TITLE
Create a standard way to pass in symbolic zero tangents and cotangent components to custom derivatives

### ISSUE THREAD

**Issue by NeilGirdhar** [google/jax#14960]
This is related to #11208.

Currently, I synthesize zero tangents like so:
```python
def zero_tangent_like(value: Array) -> RealArray:
    if jnp.issubdtype(value.dtype, jnp.inexact):
        return np.zeros_like(value)
    return np.zeros_like(value, dtype=float0)
````
But this won't work with the new Jax random keys.  I would prefer a function that works with any pytree and produces a zero tangent for it.

Lots of discussion in another issue starting here: https://github.com/google/jax/issues/14046#issuecomment-1460933659.

**Comment by mattjj** [comment]
Great suggestion.

Just to red-team the idea a bit, I think it may not be needed for `jax.jvp`, because you can already get symbolic zeros (in effect) by wrapping the function to be differentiated. That is, to get something like `jax.jvp(f, (x, y), (xdot, zero_tangent_like(y)))` we can instead write `jax.jvp(lambda x: f(x, y), (x,), (xdot,))`.

But I think this idea may be inescapable for `jax.linearize` and `jax.vjp`!

**Comment by mattjj** [comment]
> From a type annotation standpoint, would it make sense if the symbolic zero created for a Jax array also responded true to isinstance(x, jax.Array)?

That's a tricky question, and I'm going to pretend I haven't seen it yet...

**Comment by froystig** [comment]
Regarding function wrapping, we covered this a bit in discussion on #14046. It really does seem inescapable for `jax.linearize` and `jax.vjp`, and in particular @NeilGirdhar's motivating example is taking `vjp` of a function with some outputs unperturbable (new-style RNG key arrays). See https://github.com/google/jax/issues/14046#issuecomment-1460933659 and onward.

**Comment by mattjj** [comment]
Can we use `None`, as I think you suggested in that thread (but I haven't yet read it carefully)?

**Comment by mattjj** [comment]
In that other thread, Neil wrote:

> This goes beyond what's supported by None in a custom VJP backwards pass (which can only send None for an entire cotangent output—not for a component.

That sounds like something we should fix!

**Comment by NeilGirdhar** [comment]
> Can we use `None`, as I think you suggested in that thread (but I haven't yet read it carefully)?
...
> That sounds like something we should fix!

Wow, that would definitely work.

An alternative solution would be to keep Jax's behavior of disallowing `None` as components and instead create two types: a symbolic zero array that inherits from `jax.Array`, and a symbolic zero key array that inherits from `jax.random.KeyArray`.  Then, create a `jax.custom_derivatives.zero_tangent_like` with that returns the symbolic zero version of its input.

It's a bit wordier than using `None`, but the explicitness can prevent bugs where `None` is accidentally substituted.

It also means that a dataclass with, say, a `jax.Array` member wouldn't have to have `None` in that position, which is a bit nice from a type-checking standpoint.

**Comment by froystig** [comment]
Yeah, my guess is that we can use `None` (though I haven't verified), and indeed Neil made the reasonable case to allow `None` for any pytree component.

**Comment by froystig** [comment]
One thing I like about the `None` approach is that it can be made consistent with saying that we accept tree prefixes of the (co)tangent pytree. A missing subtree indicates a lack of perturbations.

**Comment by NeilGirdhar** [comment]
> One thing I like about the `None` approach is that it can be made consistent with saying that we accept tree prefixes of the (co)tangent pytree. A missing subtree indicates a lack of perturbations.

Oh, right!  Like vmap does with [its prefixes](https://jax.readthedocs.io/en/latest/pytrees.html#applying-optional-parameters-to-pytrees).  That's a nice consistency.

**Comment by NeilGirdhar** [comment]
> In that other thread, Neil wrote:
> 
> > This goes beyond what's supported by None in a custom VJP backwards pass (which can only send None for an entire cotangent output—not for a component.
> 
> That sounds like something we should fix!

In case it's helpful, there's an old issue for this: https://github.com/google/jax/issues/8641

**Comment by NeilGirdhar** [comment]
> Just to red-team the idea a bit, I think it may not be needed for `jax.jvp`, because you can already get symbolic zeros (in effect) by wrapping the function to be differentiated. That is, to get something like `jax.jvp(f, (x, y), (xdot, zero_tangent_like(y)))` we can instead write `jax.jvp(lambda x: f(x, y), (x,), (xdot,))`.

Looking back on this thread, I realize that I never answered this comment.  While it is true that this partial function application does solve the problem, there are still cases for code structure reasons why you would rather have a way of passing zeros in.  So, if it's not too difficult, it would still be nice to have an easy way to pass zeros as tangents too.  I think using `None` as you and Roy suggested is the most elegant solution.

**Comment by NeilGirdhar** [comment]
I think this is `jax.custom_derivatives.zero_from_primal` with `symbolic_zeros=True`.
